TrkReco/src/TrkRecoMcUtils_tool.cc

```diff
@@ -29,229 +29,214 @@
-    _strawDigiMCCollTag{PSet.get<std::string>("strawDigiMCCollTag")} {
-  _lastEvent = -1;
-  _mcdigis = nullptr;
-  _chColl = nullptr;
-  _mbtime = -1;
-  //    _timeOffsets = new SimParticleTimeOffset(*TimeOffsets);
-}
+  class TrkRecoMcUtils : public mu2e::McUtilsToolBase {
+    protected:
+      std::string                   _comboHitCollTag;
+      const ComboHitCollection*     _chColl;
+      std::string                   _strawDigiMCCollTag;
+      const StrawDigiMCCollection*  _mcdigis;
+      art::EventID                  _lastEvent;
+      //    SimParticleTimeOffset*       _timeOffsets;
+      double                        _mbtime;
 
-//-----------------------------------------------------------------------------
-TrkRecoMcUtils::~TrkRecoMcUtils() {
-  //    delete _timeOffsets;
-}
+    public:
+
+      TrkRecoMcUtils(const fhicl::ParameterSet& PSet);
+      ~TrkRecoMcUtils();
+
+    public:
+
+      int     initEvent(const art::Event* Event);
+
+      virtual int    strawHitSimId(const art::Event* Event, int Index) override;
 
-//-----------------------------------------------------------------------------
-int TrkRecoMcUtils::initEvent(const art::Event* Event) {
-  art::Handle<mu2e::StrawDigiMCCollection> mcdigiH;
-  Event->getByLabel(_strawDigiMCCollTag, mcdigiH);
-  if (mcdigiH.isValid())
-    _mcdigis = (mu2e::StrawDigiMCCollection*)mcdigiH.product();
-  else
-    _mcdigis = NULL;
-
-  auto chcH = Event->getValidHandle<ComboHitCollection>(_comboHitCollTag);
-  _chColl = chcH.product();
-
-  //    _timeOffsets->updateMap(*Event);
-  if (_mbtime < 0) {
-    ConditionsHandle<AcceleratorParams> accPar("ignored");
-    _mbtime = accPar->deBuncherPeriod;
+      virtual double mcDoca(const art::Event* Event, const TrkStrawHit* StrawHit) override;
+      // int               Index,
+      // const Straw*      Straw) override ;
+
+      // virtual int    nGenHits(const art::Event*         Event      ,
+      //          fhicl::ParameterSet*      TimeOffsets,
+      //          const StrawHitCollection* Shcol      ) override;
+
+      // virtual const StrawDigiMCCollection* getListOfMcStrawHits(const art::Event* Event,
+      //                    const art::InputTag& Tag) override;
+
+      virtual const SimParticle* getSimParticle(const art::Event* Event, int HitIndex) override;
+      virtual const XYZVectorF* getMom(const art::Event* Event, int HitIndex) override;
+
+      int   getID      (const SimParticle* Sim) override;
+      int   getPdgID   (const SimParticle* Sim) override;
+      float getStartMom(const SimParticle* Sim) override;
+
+  };
+
+  //-----------------------------------------------------------------------------
+  TrkRecoMcUtils::TrkRecoMcUtils(const fhicl::ParameterSet& PSet) :
+    _comboHitCollTag   { PSet.get<std::string>("comboHitCollTag"   ) },
+    _strawDigiMCCollTag{ PSet.get<std::string>("strawDigiMCCollTag") }
+  {
+    _lastEvent   = art::EventID(0,0,0);
```

Comment:
This will compile but will cause a run time error.  Just default construct it.  Internally the art::EventID has a reserved values for each field meaning "invalid".   When you default construct it, each field gets it's invalid value.  No legal event ID can have any field invalid.  So they have already covered this case for you.